File: api/license_key.go
package api

import (
	"encoding/json"
	"time"

	"github.com/distr-sh/distr/internal/types"
	"github.com/google/uuid"
)

type CreateLicenseKeyRequest struct {
	Name                   string          `json:"name"`
	Description            *string         `json:"description,omitempty"`
	Payload                json.RawMessage `json:"payload,omitempty"`
	NotBefore              time.Time       `json:"notBefore,omitempty"`
	ExpiresAt              time.Time       `json:"expiresAt,omitempty"`
	CustomerOrganizationID *uuid.UUID      `json:"customerOrganizationId,omitempty"`
	LicenseTemplateID      *uuid.UUID      `json:"licenseTemplateId,omitempty"`
}

type UpdateLicenseKeyRequest struct {
	Description       *string          `json:"description,omitempty"`
	NotBefore         *time.Time       `json:"notBefore,omitempty"`
	ExpiresAt         *time.Time       `json:"expiresAt,omitempty"`
	Payload           *json.RawMessage `json:"payload,omitempty"`
	LicenseTemplateID *uuid.UUID       `json:"licenseTemplateId,omitempty"`
	Confirm           bool             `query:"confirm"`
}

type LicenseKeyRevision struct {
	types.LicenseKeyRevision

	Token string `json:"token"`
}

type UpdateLicenseKeyResponse struct {
	types.LicenseKey
	AffectedDeployments []AffectedDeployment `json:"affectedDeployments"`
}

File: frontend/ui/src/app/types/license-key.ts
import {BaseModel, Named} from '@distr-sh/distr-sdk';

export interface LicenseKey extends BaseModel, Named {
  description?: string;
  payload?: Record<string, unknown>;
  notBefore?: string;
  expiresAt?: string;
  lastRevisedAt?: string;
  customerOrganizationId?: string;
  licenseTemplateId?: string;
}

export interface CreateLicenseKeyRequest {
  name: string;
  description?: string;
  payload?: Record<string, unknown>;
  notBefore?: string;
  expiresAt?: string;
  customerOrganizationId?: string;
  licenseTemplateId?: string;
}

export interface UpdateLicenseKeyRequest {
  description?: string;
  payload?: Record<string, unknown>;
  notBefore?: string;
  expiresAt?: string;
  licenseTemplateId?: string;
}

export interface LicenseKeyRevision extends BaseModel {
  notBefore: string;
  expiresAt: string;
  payload: Record<string, unknown>;
  token: string;
}
